<h3 style='color:blue' align='center'>Handling imbalanced data in customer churn prediction</h3>

Customer churn prediction is to measure why customers are leaving a business. In this tutorial we will be looking at customer churn in telecom business. We will build a deep learning model to predict the churn and use precision,recall, f1-score to measure performance of our model.
We will then handle imbalance in data using various techniques and improve f1-score

In [1]:
import pandas as pd
from matplotlib import pyplot as plt
import numpy as np
%matplotlib inline

In [2]:
import warnings
warnings.filterwarnings('ignore')

**Load the data**

In [3]:
df1 = pd.read_csv("random_imputed.csv",index_col=[0])
df1.sample(5)

gender  body.temperature  pulse  respiration  \
inpatient.number                                                 
835982                 1              36.0    100           19   
742228                 0              36.3     66           19   
775827                 0              36.2     66           18   
736744                 1              37.1    121           19   
840048                 1              36.5     76           21   

                  systolic.blood.pressure  diastolic.blood.pressure  \
inpatient.number                                                      
835982                                140                        72   
742228                                 90                        68   
775827                                130                        80   
736744                                162                       100   
840048                                120                        60   

                         map        BMI  type.of.heart.failure  \
inpatient.number                                                 
835982             94.666667  19.723866                      2   
742228             75.333333  20.957171                      2   
775827             96.666667  23.555556                      2   
736744            120.666667  27.886696                      0   
840048             80.000000  18.442546                      2   

                  NYHA.cardiac.function.classification  ...  \
inpatient.number                                        ...   
835982                                               3  ...   
742228                                               3  ...   
775827                                               4  ...   
736744                                               3  ...   
840048                                               3  ...   

                  measured.bicarbonate  carboxyhemoglobin  oxygen.saturation  \
inpatient.number                                                               
835982                       23.246503           1.217008          96.585423   
742228                       26.900000           0.000000          98.000000   
775827                       23.246503           1.217008          96.585423   
736744                       23.246503           1.217008          96.585423   
840048                       27.000000           0.300000          98.000000   

                  partial.oxygen.pressure  oxyhemoglobin  anion.gap  \
inpatient.number                                                      
835982                           86.74564      94.749285  15.874385   
742228                          121.00000      97.700000   8.400000   
775827                           86.74564      94.749285  15.874385   
736744                           86.74564      94.749285  15.874385   
840048                          109.00000      97.700000  18.000000   

                  free.calcium  total.hemoglobin  GCS  ageCat  
inpatient.number                                               
835982                1.252327        151.929679   15      75  
742228                1.000000         92.000000   15      85  
775827                1.252327        151.929679   15      75  
736744                1.252327        151.929679   15      55  
840048                1.020000         93.000000   15      85  

[5 rows x 128 columns]

In [4]:
df3=pd.read_csv("drug_onehot_latest.csv",index_col=[0])
df3.head()


sulfotanshinone sodium injection  Furosemide tablet  \
inpatient.number                                                        
722128                                           0                  1   
723327                                           0                  1   
723617                                           0                  0   
724385                                           0                  1   
725509                                           0                  1   

                  Meglumine Adenosine Cyclophosphate for injection  \
inpatient.number                                                     
722128                                                           0   
723327                                                           0   
723617                                                           0   
724385                                                           0   
725509                                                           0   

                  Furosemide injection  Milrinone injection  \
inpatient.number                                              
722128                               1                    1   
723327                               1                    1   
723617                               1                    0   
724385                               1                    1   
725509                               1                    1   

                  Deslanoside injection  Torasemide tablet  \
inpatient.number                                             
722128                                1                  0   
723327                                1                  1   
723617                                0                  0   
724385                                1                  0   
725509                                1                  1   

                  Benazepril hydrochloride tablet  \
inpatient.number                                    
722128                                          0   
723327                                          0   
723617                                          0   
724385                                          0   
725509                                          0   

                  Atorvastatin calcium tablet  Digoxin tablet  \
inpatient.number                                                
722128                                      0               1   
723327                                      1               1   
723617                                      1               0   
724385                                      0               1   
725509                                      0               1   

                  Hydrochlorothiazide tablet  Spironolactone tablet  \
inpatient.number                                                      
722128                                     0                      1   
723327                                     0                      1   
723617                                     0                      1   
724385                                     0                      1   
725509                                     0                      1   

                  Valsartan Dispersible tablet  \
inpatient.number                                 
722128                                       0   
723327                                       1   
723617                                       0   
724385                                       0   
725509                                       0   

                  Dobutamine hydrochloride injection  \
inpatient.number                                       
722128                                             0   
723327                                             0   
723617                                             0   
724385                                             0   
725509                                             0   

                  Isoprenaline Hydrochloride injection  \
inpatient.

In [6]:
df2=df3['Atorvastatin calcium tablet']
df2 = df2.to_frame()

In [7]:
df2.head()

Atorvastatin calcium tablet
inpatient.number                             
722128                                      0
723327                                      1
723617                                      1
724385                                      0
725509                                      0

In [8]:
df=pd.concat([df1,df2],axis=1)

In [9]:
df.head()

gender  body.temperature  pulse  respiration  \
inpatient.number                                                 
722128                 0              36.9     66           18   
723327                 0              36.0     72           21   
723617                 1              36.4     70           20   
724385                 1              36.7    129           22   
725509                 0              36.9    100           20   

                  systolic.blood.pressure  diastolic.blood.pressure  \
inpatient.number                                                      
722128                                121                        56   
723327                                150                        80   
723617                                120                        80   
724385                                122                        86   
725509                                 95                        65   

                         map        BMI  type.of.heart.failure  \
inpatient.number                                                 
722128             77.666667  18.590125                      2   
723327            103.333333  21.644121                      2   
723617             93.333333  31.111111                      2   
724385             98.000000  20.134780                      2   
725509             75.000000  22.031726                      2   

                  NYHA.cardiac.function.classification  ...  \
inpatient.number                                        ...   
722128                                               4  ...   
723327                                               4  ...   
723617                                               3  ...   
724385                                               3  ...   
725509                                               3  ...   

                  carboxyhemoglobin  oxygen.saturation  \
inpatient.number                                         
722128                     1.217008          96.585423   
723327                     0.700000          91.000000   
723617                     0.000000          98.000000   
724385                     1.217008          96.585423   
725509                     1.217008          96.585423   

                  partial.oxygen.pressure  oxyhemoglobin  anion.gap  \
inpatient.number                                                      
722128                           86.74564      94.749285  15.874385   
723327                           64.00000      90.000000  13.000000   
723617                          106.00000      97.700000   9.100000   
724385                           86.74564      94.749285  15.874385   
725509                           86.74564      94.749285  15.874385   

                  free.calcium  total.hemoglobin  GCS  ageCat  \
inpatient.number                                                
722128                1.252327        151.929679   15      75   
723327                1.110000         73.000000   15      65   
723617                1.110000        104.000000   15      85   
724385                1.252327        151.929679   15      75   
725509                1.252327        151.929679   15      45   

                  Atorvastatin calcium tablet  
inpatient.number                               
722128                                      0  
723327                                      1  
723617                                      1  
724385                                      0  
725509                                      0  

[5 rows x 129 columns]

In [10]:
517400/df.shape[0]

258.3125312031952

**First of all, drop customerID column as it is of no use**

In [11]:
#df.drop('customerID',axis='columns',inplace=True)

In [12]:
df.dtypes

gender                           int64
body.temperature               float64
pulse                            int64
respiration                      int64
systolic.blood.pressure          int64
                                ...   
free.calcium                   float64
total.hemoglobin               float64
GCS                              int64
ageCat                           int64
Atorvastatin calcium tablet      int64
Length: 129, dtype: object

**Quick glance at above makes me realize that TotalCharges should be float but it is an object. Let's check what's going on with  this column**

In [13]:
#df.TotalCharges.values

**Ahh... it is string. Lets convert it to numbers**

In [14]:
#pd.to_numeric(df.TotalCharges,errors='coerce').isnull()

In [15]:
#df[pd.to_numeric(df.TotalCharges,errors='coerce').isnull()]

In [16]:
df.shape

(2003, 129)

In [17]:
#df.iloc[488].TotalCharges

In [18]:
#df[df.TotalCharges!=' '].shape

**Remove rows with space in TotalCharges**

In [19]:
#df1 = df[df.TotalCharges!=' ']
#df1.shape

In [20]:
# df1.dtypes

In [21]:
# df1.TotalCharges = pd.to_numeric(df1.TotalCharges)

In [22]:
# df1.TotalCharges.values

In [23]:
# df1[df1.Churn=='No']

**Data Visualization**

In [24]:
# tenure_churn_no = df1[df1.Churn=='No'].tenure
# tenure_churn_yes = df1[df1.Churn=='Yes'].tenure

# plt.xlabel("tenure")
# plt.ylabel("Number Of Customers")
# plt.title("Customer Churn Prediction Visualiztion")

# blood_sugar_men = [113, 85, 90, 150, 149, 88, 93, 115, 135, 80, 77, 82, 129]
# blood_sugar_women = [67, 98, 89, 120, 133, 150, 84, 69, 89, 79, 120, 112, 100]

# plt.hist([tenure_churn_yes, tenure_churn_no], rwidth=0.95, color=['green','red'],label=['Churn=Yes','Churn=No'])
# plt.legend()

In [25]:
# mc_churn_no = df1[df1.Churn=='No'].MonthlyCharges      
# mc_churn_yes = df1[df1.Churn=='Yes'].MonthlyCharges      

# plt.xlabel("Monthly Charges")
# plt.ylabel("Number Of Customers")
# plt.title("Customer Churn Prediction Visualiztion")

# blood_sugar_men = [113, 85, 90, 150, 149, 88, 93, 115, 135, 80, 77, 82, 129]
# blood_sugar_women = [67, 98, 89, 120, 133, 150, 84, 69, 89, 79, 120, 112, 100]

# plt.hist([mc_churn_yes, mc_churn_no], rwidth=0.95, color=['green','red'],label=['Churn=Yes','Churn=No'])
# plt.legend()

**Many of the columns are yes, no etc. Let's print unique values in object columns to see data values**

In [26]:
# # def print_unique_col_values(df):
#        for column in df:
#             if df[column].dtypes=='object':
#                 print(f'{column}: {df[column].unique()}') 

In [27]:
# print_unique_col_values(df2)

**Some of the columns have no internet service or no phone service, that can be replaced with a simple No**

In [28]:
# df1.replace('No internet service','No',inplace=True)
# df1.replace('No phone service','No',inplace=True)

In [29]:
# print_unique_col_values(df1)

**Convert Yes and No to 1 or 0**

In [30]:
# yes_no_columns = ['Partner','Dependents','PhoneService','MultipleLines','OnlineSecurity','OnlineBackup',
#                   'DeviceProtection','TechSupport','StreamingTV','StreamingMovies','PaperlessBilling','Churn']
# for col in yes_no_columns:
#     df1[col].replace({'Yes': 1,'No': 0},inplace=True)

In [31]:
# for col in df1:
#     print(f'{col}: {df1[col].unique()}') 

In [32]:
#df1['gender'].replace({'Female':1,'Male':0},inplace=True)

In [33]:
# df1.gender.unique()

**One hot encoding for categorical columns**

In [34]:
# df2 = pd.get_dummies(data=df1, columns=['InternetService','Contract','PaymentMethod'])
# df2.columns

In [35]:
# df2.sample(5)

In [36]:
# df2.dtypes

In [37]:
cols_to_scale = [df.columns]

from sklearn.preprocessing import MinMaxScaler
scaler = MinMaxScaler()
scaled_df= scaler.fit_transform(df)

In [38]:
df = pd.DataFrame(scaled_df,columns = df.columns)

In [39]:
df.head()

gender  body.temperature     pulse  respiration  systolic.blood.pressure  \
0     0.0          0.271429  0.333333     0.500000                 0.480159   
1     0.0          0.142857  0.363636     0.583333                 0.595238   
2     1.0          0.200000  0.353535     0.555556                 0.476190   
3     1.0          0.242857  0.651515     0.611111                 0.484127   
4     0.0          0.271429  0.505051     0.555556                 0.376984   

   diastolic.blood.pressure       map       BMI  type.of.heart.failure  \
0                  0.383562  0.428309  0.046006                    1.0   
1                  0.547945  0.569853  0.053564                    1.0   
2                  0.547945  0.514706  0.076992                    1.0   
3                  0.589041  0.540441  0.049828                    1.0   
4                  0.445205  0.413603  0.054523                    1.0   

   NYHA.cardiac.function.classification  ...  carboxyhemoglobin  \
0                                   1.0  ...           0.229624   
1                                   1.0  ...           0.132075   
2                                   0.5  ...           0.000000   
3                                   0.5  ...           0.229624   
4                                   0.5  ...           0.229624   

   oxygen.saturation  partial.oxygen.pressure  oxyhemoglobin  anion.gap  \
0           0.954472                 0.284024       0.941835   0.380276   
1           0.880000                 0.187234       0.878342   0.316258   
2           0.973333                 0.365957       0.981283   0.229399   
3           0.954472                 0.284024       0.941835   0.380276   
4           0.954472                 0.284024       0.941835   0.380276   

   free.calcium  total.hemoglobin  GCS    ageCat  Atorvastatin calcium tablet  
0      0.724655          0.696010  1.0  0.666667                          0.0  
1      0.440000          0.226190  1.0  0.533333                          1.0  
2      0.440000          0.410714  1.0  0.800000                          1.0  
3      0.724655          0.696010  1.0  0.666667                          0.0  
4      0.724655          0.696010  1.0  0.266667                          0.0  

[5 rows x 129 columns]

**Train test split**

In [40]:
X = df2.drop('Atorvastatin calcium tablet',axis='columns')
y = testLabels = df2['Atorvastatin calcium tablet'].astype(np.float32)

from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(df, df2, test_size=0.2, random_state=15, stratify=df2)

In [41]:
y_train.value_counts()

Atorvastatin calcium tablet
0                              948
1                              654
dtype: int64

In [42]:
y.value_counts()

0.0    1185
1.0     818
Name: Atorvastatin calcium tablet, dtype: int64

In [43]:
1185/818

1.4486552567237163

In [44]:
y_test.value_counts()

Atorvastatin calcium tablet
0                              237
1                              164
dtype: int64

In [45]:
X_train.shape

(1602, 129)

In [46]:
X_test.shape

(401, 129)

In [47]:
X_train[:10]

gender  body.temperature     pulse  respiration  \
1486     0.0          0.185714  0.494949     0.555556   
223      1.0          0.228571  0.353535     0.500000   
1034     0.0          0.214286  0.333333     0.500000   
762      0.0          0.142857  0.313131     0.500000   
188      1.0          0.242857  0.303030     0.527778   
1909     1.0          0.285714  0.328283     0.500000   
1087     1.0          0.142857  0.520202     0.555556   
1955     0.0          0.257143  0.419192     0.500000   
606      0.0          0.371429  0.414141     0.555556   
526      0.0          0.228571  0.338384     0.500000   

      systolic.blood.pressure  diastolic.blood.pressure       map       BMI  \
1486                 0.714286                  0.616438  0.661765  0.054994   
223                  0.357143                  0.410959  0.386029  0.048805   
1034                 0.492063                  0.479452  0.485294  0.043263   
762                  0.388889                  0.458904  0.426471  0.046353   
188                  0.531746                  0.616438  0.577206  0.063431   
1909                 0.595238                  0.479452  0.533088  0.045641   
1087                 0.436508                  0.547945  0.496324  0.066357   
1955                 0.543651                  0.568493  0.556985  0.052795   
606                  0.484127                  0.479452  0.481618  0.047097   
526                  0.515873                  0.465753  0.488971  0.045761   

      type.of.heart.failure  NYHA.cardiac.function.classification  ...  \
1486                    1.0                                   0.0  ...   
223                     1.0                                   1.0  ...   
1034                    0.0                                   0.5  ...   
762                     1.0                                   0.5  ...   
188                     1.0                                   0.5  ...   
1909                    0.0                                   0.0  ...   
1087                    1.0                                   0.5  ...   
1955                    1.0                                   0.5  ...   
606                     0.0                                   0.5  ...   
526                     1.0                                   1.0  ...   

      carboxyhemoglobin  oxygen.saturation  partial.oxygen.pressure  \
1486           0.094340           0.986667                 0.472340   
223            0.056604           0.986667                 0.506383   
1034           0.018868           0.986667                 0.710638   
762            0.094340           0.946667                 0.234043   
188            0.229624           0.954472                 0.284024   
1909           0.229624           0.954472                 0.284024   
1087           0.056604           0.986667                 0.544681   
1955           0.229624           0.954472                 0.284024   
606            0.229624           0.954472                 0.284024   
526            0.283019           0.640000                 0.102128   

      oxyhemoglobin  anion.gap  free.calcium  total.hemoglobin       GCS  \
1486       0.982620   0.340757      0.440000          0.583333  1.000000   
223        0.987968   0.296214      0.500000          0.619048  1.000000   
1034       1.000000   0.354120      0.520000          0.547619  1.000000   
762        0.943850   0.400891      0.320000          0.726190  1.000000   
188        0.941835   0.380276      0.724655          0.696010  0.583333   
1909       0.941835   0.380276      0.724655          0.696010  1.000000   
1087       0.989305   0.191537      0.320000          0.428571  1.000000   
1955       0.941835   0.380276      0.724655          0.696010  1.000000   
606        0.941835   0.380276      0.724655          0.696010  1.000000   
526        0.631016   0.360802      0.360000          0.470238  1.000000   

        ageCat  Atorvastatin calcium tablet  
1486  0.800000              

In [48]:
# len(X_train.columns)

**Build a model (ANN) in tensorflow/keras**

In [262]:
from tensorflow_addons import losses

ModuleNotFoundError: No module named 'tensorflow_addons'

In [49]:
import tensorflow as tf
from tensorflow import keras
from sklearn.metrics import confusion_matrix , classification_report

In [73]:
def ANN(X_train, y_train, X_test, y_test, loss, weights):
    model = keras.Sequential([
        keras.layers.Dense(128, input_dim=128, activation='relu'),
        keras.layers.Dense(15, activation='relu'),
        keras.layers.Dense(1, activation='sigmoid')
    ])

    model.compile(optimizer='rmsprop', loss=loss, metrics=['accuracy'])
    
    if weights == -1:
        model.fit(X_train, y_train, epochs=50)
    else:
        model.fit(X_train, y_train, epochs=50, class_weight = weights)
    
    print(model.evaluate(X_test, y_test))
    
    y_preds = model.predict(X_test)
    y_preds = np.round(y_preds)
    
    print("Classification Report: \n", classification_report(y_test, y_preds))
    
    return y_preds

In [66]:
y_preds = ANN(X_train, y_train, X_test, y_test, 'binary_crossentropy', -1)

Epoch 1/50
60/60 [==============================] - 0s 1ms/step - loss: 0.6989 - accuracy: 0.5222
Epoch 2/50
60/60 [==============================] - 0s 1ms/step - loss: 0.6725 - accuracy: 0.5939
Epoch 3/50
60/60 [==============================] - 0s 1ms/step - loss: 0.6617 - accuracy: 0.5997
Epoch 4/50
60/60 [==============================] - 0s 1ms/step - loss: 0.6491 - accuracy: 0.6282
Epoch 5/50
60/60 [==============================] - 0s 1ms/step - loss: 0.6405 - accuracy: 0.6535
Epoch 6/50
60/60 [==============================] - 0s 1ms/step - loss: 0.6377 - accuracy: 0.6340
Epoch 7/50
60/60 [==============================] - 0s 1ms/step - loss: 0.6293 - accuracy: 0.6466
Epoch 8/50
60/60 [==============================] - 0s 1ms/step - loss: 0.6251 - accuracy: 0.6487
Epoch 9/50
60/60 [==============================] - 0s 2ms/step - loss: 0.6169 - accuracy: 0.6609
Epoch 10/50
60/60 [==============================] - 0s 1ms/step - loss: 0.6155 - accuracy: 0.6624
Epoch 11/50
60/60 [

## Mitigating Skewdness of Data

### Method 1: Undersampling

reference: https://www.kaggle.com/rafjaa/resampling-strategies-for-imbalanced-datasets

In [51]:
# Class count
count_class_0, count_class_1 = df['Atorvastatin calcium tablet'].value_counts()

# Divide by class
df_class_0 = df[df['Atorvastatin calcium tablet'] == 0]
df_class_1 = df[df['Atorvastatin calcium tablet'] == 1]

In [52]:
df['Atorvastatin calcium tablet'].value_counts()

0.0    1185
1.0     818
Name: Atorvastatin calcium tablet, dtype: int64

In [53]:
# Undersample 0-class and concat the DataFrames of both class
df_class_0_under = df_class_0.sample(count_class_1)
df_test_under = pd.concat([df_class_0_under, df_class_1], axis=0)

print('Random under-sampling:')
print(df_test_under['Atorvastatin calcium tablet'].value_counts())

Random under-sampling:
0.0    818
1.0    818
Name: Atorvastatin calcium tablet, dtype: int64


In [54]:
X = df_test_under.drop('Atorvastatin calcium tablet',axis='columns')
y = df_test_under['Atorvastatin calcium tablet']

from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=15, stratify=y)

In [55]:
# Number of classes in training Data
y_train.value_counts()

0.0    654
1.0    654
Name: Atorvastatin calcium tablet, dtype: int64

**Printing Classification in the last, Scroll down till the last epoch to watch the classification report**

In [274]:
y_preds = ANN(X_train, y_train, X_test, y_test, 'binary_crossentropy', -1)

Epoch 1/20
41/41 [==============================] - 1s 4ms/step - loss: 0.6893 - accuracy: 0.5252
Epoch 2/20
41/41 [==============================] - 0s 2ms/step - loss: 0.6817 - accuracy: 0.5635
Epoch 3/20
41/41 [==============================] - 0s 1ms/step - loss: 0.6726 - accuracy: 0.6047
Epoch 4/20
41/41 [==============================] - 0s 1ms/step - loss: 0.6626 - accuracy: 0.6177
Epoch 5/20
41/41 [==============================] - 0s 1ms/step - loss: 0.6564 - accuracy: 0.6009
Epoch 6/20
41/41 [==============================] - 0s 1ms/step - loss: 0.6517 - accuracy: 0.6131
Epoch 7/20
41/41 [==============================] - 0s 1ms/step - loss: 0.6405 - accuracy: 0.6323
Epoch 8/20
41/41 [==============================] - 0s 1ms/step - loss: 0.6362 - accuracy: 0.6460
Epoch 9/20
41/41 [==============================] - 0s 1ms/step - loss: 0.6329 - accuracy: 0.6323
Epoch 10/20
41/41 [==============================] - 0s 1ms/step - loss: 0.6261 - accuracy: 0.6437
Epoch 11/20
41/41 [

Check classification report above. f1-score for minority class 1 improved from **0.57 to 0.76**. Score for class 0 reduced to 0.75 from 0.85 but that's ok. We have more generalized classifier which classifies both classes with similar prediction score

### Method2: Oversampling

In [57]:
# Oversample 1-class and concat the DataFrames of both classes
df_class_1_over = df_class_1.sample(count_class_0, replace=True)
df_test_over = pd.concat([df_class_0, df_class_1_over], axis=0)

print('Random over-sampling:')
print(df_test_over['Atorvastatin calcium tablet'].value_counts())

Random over-sampling:
0.0    1185
1.0    1185
Name: Atorvastatin calcium tablet, dtype: int64


In [58]:
X = df_test_over.drop('Atorvastatin calcium tablet',axis='columns')
y = df_test_over['Atorvastatin calcium tablet']

from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=15, stratify=y)

In [59]:
# Number of classes in training Data
y_train.value_counts()

0.0    948
1.0    948
Name: Atorvastatin calcium tablet, dtype: int64

In [74]:
loss = keras.losses.BinaryCrossentropy()
weights = -1
y_preds = ANN(X_train, y_train, X_test, y_test, 'binary_crossentropy', -1)

Epoch 1/50
60/60 [==============================] - 1s 1ms/step - loss: 0.6827 - accuracy: 0.5649
Epoch 2/50
60/60 [==============================] - 0s 2ms/step - loss: 0.6654 - accuracy: 0.6044
Epoch 3/50
60/60 [==============================] - 0s 1ms/step - loss: 0.6535 - accuracy: 0.6213
Epoch 4/50
60/60 [==============================] - 0s 1ms/step - loss: 0.6415 - accuracy: 0.6276
Epoch 5/50
60/60 [==============================] - 0s 1ms/step - loss: 0.6328 - accuracy: 0.6530
Epoch 6/50
60/60 [==============================] - 0s 1ms/step - loss: 0.6340 - accuracy: 0.6355
Epoch 7/50
60/60 [==============================] - 0s 2ms/step - loss: 0.6238 - accuracy: 0.6524
Epoch 8/50
60/60 [==============================] - 0s 2ms/step - loss: 0.6173 - accuracy: 0.6508
Epoch 9/50
60/60 [==============================] - 0s 2ms/step - loss: 0.6161 - accuracy: 0.6582
Epoch 10/50
60/60 [==============================] - 0s 2ms/step - loss: 0.6086 - accuracy: 0.6640
Epoch 11/50
60/60 [

Check classification report above. f1-score for minority class 1 improved from **0.57 to 0.76**. Score for class 0 reduced to 0.75 from 0.85 but that's ok. We have more generalized classifier which classifies both classes with similar prediction score

### Method3: SMOTE

To install imbalanced-learn library use **pip install imbalanced-learn** command

In [71]:
X = df.drop('Atorvastatin calcium tablet',axis='columns')
y = df['Atorvastatin calcium tablet']

In [61]:
from imblearn.over_sampling import SMOTE

smote = SMOTE(sampling_strategy='minority')
X_sm, y_sm = smote.fit_sample(X, y)

y_sm.value_counts()

1    5163
0    5163
Name: Churn, dtype: int64

In [62]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X_sm, y_sm, test_size=0.2, random_state=15, stratify=y_sm)

In [63]:
# Number of classes in training Data
y_train.value_counts()

1    4130
0    4130
Name: Churn, dtype: int64

In [64]:
y_preds = ANN(X_train, y_train, X_test, y_test, 'binary_crossentropy', -1)

Epoch 1/100
259/259 [==============================] - 0s 1ms/step - loss: 0.5222 - accuracy: 0.7470
Epoch 2/100
259/259 [==============================] - 0s 1ms/step - loss: 0.4716 - accuracy: 0.7712
Epoch 3/100
259/259 [==============================] - 0s 1ms/step - loss: 0.4627 - accuracy: 0.7798
Epoch 4/100
259/259 [==============================] - 0s 1ms/step - loss: 0.4552 - accuracy: 0.7824
Epoch 5/100
259/259 [==============================] - 0s 1ms/step - loss: 0.4518 - accuracy: 0.7852
Epoch 6/100
259/259 [==============================] - 0s 1ms/step - loss: 0.4472 - accuracy: 0.7860
Epoch 7/100
259/259 [==============================] - 0s 1ms/step - loss: 0.4443 - accuracy: 0.7904
Epoch 8/100
259/259 [==============================] - 0s 1ms/step - loss: 0.4396 - accuracy: 0.7953
Epoch 9/100
259/259 [==============================] - 0s 1ms/step - loss: 0.4367 - accuracy: 0.7944
Epoch 10/100
259/259 [==============================] - 0s 1ms/step - loss: 0.4350 - accura

SMOT Oversampling increases f1 score of minority class 1 from **0.57 to 0.81 (huge improvement)** Also over all accuracy improves from 0.78 to 0.80

### Method4: Use of Ensemble with undersampling

In [65]:
df2.Churn.value_counts()

0    5163
1    1869
Name: Churn, dtype: int64

In [66]:
# Regain Original features and labels
X = df2.drop('Churn',axis='columns')
y = df2['Churn']

In [67]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=15, stratify=y)

In [68]:
y_train.value_counts()

0    4130
1    1495
Name: Churn, dtype: int64

model1 --> class1(1495) + class0(0, 1495)

model2 --> class1(1495) + class0(1496, 2990)

model3 --> class1(1495) + class0(2990, 4130)

In [73]:
df3 = X_train.copy()
df3['Churn'] = y_train

In [74]:
df3.head()

gender  SeniorCitizen  Partner  Dependents    tenure  PhoneService  \
684        1              0        0           0  0.000000             1   
2446       1              0        0           0  0.239437             1   
1680       0              0        1           1  0.774648             1   
2220       0              0        1           0  1.000000             1   
2842       1              0        0           0  0.042254             0   

      MultipleLines  OnlineSecurity  OnlineBackup  DeviceProtection  ...  \
684               0               0             0                 0  ...   
2446              1               0             1                 0  ...   
1680              1               0             0                 0  ...   
2220              0               1             1                 0  ...   
2842              0               1             0                 1  ...   

      InternetService_Fiber optic  InternetService_No  \
684                             1                   0   
2446                            1                   0   
1680                            0                   1   
2220                            0                   0   
2842                            0                   0   

      Contract_Month-to-month  Contract_One year  Contract_Two year  \
684                         1                  0                  0   
2446                        1                  0                  0   
1680                        0                  1                  0   
2220                        0                  0                  1   
2842                        1                  0                  0   

      PaymentMethod_Bank transfer (automatic)  \
684                                         0   
2446                                        0   
1680                                        0   
2220                                        1   
2842                                        0   

      PaymentMethod_Credit card (automatic)  PaymentMethod_Electronic check  \
684                                       0                               0   
2446                                      1                               0   
1680                                      0                               0   
2220                                      0                               0   
2842                                      0                               0   

      PaymentMethod_Mailed check  Churn  
684                            1      0  
2446                           0      1  
1680                           1      0  
2220                           0      0  
2842                           1      0  

[5 rows x 27 columns]

In [75]:
df3_class0 = df3[df3.Churn==0]
df3_class1 = df3[df3.Churn==1]

In [76]:
def get_train_batch(df_majority, df_minority, start, end):
    df_train = pd.concat([df_majority[start:end], df_minority], axis=0)

    X_train = df_train.drop('Churn', axis='columns')
    y_train = df_train.Churn
    return X_train, y_train    

In [77]:
X_train, y_train = get_train_batch(df3_class0, df3_class1, 0, 1495)

y_pred1 = ANN(X_train, y_train, X_test, y_test, 'binary_crossentropy', -1)

Epoch 1/100
94/94 [==============================] - 0s 1ms/step - loss: 0.5840 - accuracy: 0.7164
Epoch 2/100
94/94 [==============================] - 0s 1ms/step - loss: 0.5094 - accuracy: 0.7542
Epoch 3/100
94/94 [==============================] - 0s 1ms/step - loss: 0.4964 - accuracy: 0.7656
Epoch 4/100
94/94 [==============================] - 0s 1ms/step - loss: 0.4893 - accuracy: 0.7649
Epoch 5/100
94/94 [==============================] - 0s 1ms/step - loss: 0.4844 - accuracy: 0.7672
Epoch 6/100
94/94 [==============================] - 0s 1ms/step - loss: 0.4817 - accuracy: 0.7689
Epoch 7/100
94/94 [==============================] - 0s 1ms/step - loss: 0.4767 - accuracy: 0.7719
Epoch 8/100
94/94 [==============================] - 0s 1ms/step - loss: 0.4758 - accuracy: 0.7722
Epoch 9/100
94/94 [==============================] - 0s 1ms/step - loss: 0.4733 - accuracy: 0.7753
Epoch 10/100
94/94 [==============================] - 0s 1ms/step - loss: 0.4702 - accuracy: 0.7753
Epoch 11/

94/94 [==============================] - 0s 1ms/step - loss: 0.3921 - accuracy: 0.8244
Epoch 84/100
94/94 [==============================] - 0s 1ms/step - loss: 0.3879 - accuracy: 0.8247
Epoch 85/100
94/94 [==============================] - 0s 1ms/step - loss: 0.3908 - accuracy: 0.8284
Epoch 86/100
94/94 [==============================] - 0s 1ms/step - loss: 0.3880 - accuracy: 0.8221
Epoch 87/100
94/94 [==============================] - 0s 2ms/step - loss: 0.3894 - accuracy: 0.8258
Epoch 88/100
94/94 [==============================] - 0s 1ms/step - loss: 0.3862 - accuracy: 0.8271
Epoch 89/100
94/94 [==============================] - 0s 1ms/step - loss: 0.3881 - accuracy: 0.8284
Epoch 90/100
94/94 [==============================] - 0s 1ms/step - loss: 0.3843 - accuracy: 0.8268
Epoch 91/100
94/94 [==============================] - 0s 1ms/step - loss: 0.3847 - accuracy: 0.8268
Epoch 92/100
94/94 [==============================] - 0s 1ms/step - loss: 0.3865 - accuracy: 0.8268
Epoch 93/100


In [78]:
X_train, y_train = get_train_batch(df3_class0, df3_class1, 1495, 2990)

y_pred2 = ANN(X_train, y_train, X_test, y_test, 'binary_crossentropy', -1)

Epoch 1/100
94/94 [==============================] - 0s 1ms/step - loss: 0.6486 - accuracy: 0.6328
Epoch 2/100
94/94 [==============================] - 0s 1ms/step - loss: 0.5236 - accuracy: 0.7522
Epoch 3/100
94/94 [==============================] - 0s 1ms/step - loss: 0.5005 - accuracy: 0.7572
Epoch 4/100
94/94 [==============================] - 0s 1ms/step - loss: 0.4935 - accuracy: 0.7592
Epoch 5/100
94/94 [==============================] - 0s 1ms/step - loss: 0.4883 - accuracy: 0.7562
Epoch 6/100
94/94 [==============================] - 0s 2ms/step - loss: 0.4845 - accuracy: 0.7605
Epoch 7/100
94/94 [==============================] - 0s 2ms/step - loss: 0.4795 - accuracy: 0.7625
Epoch 8/100
94/94 [==============================] - 0s 1ms/step - loss: 0.4778 - accuracy: 0.7642
Epoch 9/100
94/94 [==============================] - 0s 1ms/step - loss: 0.4747 - accuracy: 0.7659
Epoch 10/100
94/94 [==============================] - 0s 1ms/step - loss: 0.4719 - accuracy: 0.7692
Epoch 11/

94/94 [==============================] - 0s 1ms/step - loss: 0.3779 - accuracy: 0.8227
Epoch 84/100
94/94 [==============================] - 0s 1ms/step - loss: 0.3769 - accuracy: 0.8221
Epoch 85/100
94/94 [==============================] - 0s 1ms/step - loss: 0.3758 - accuracy: 0.8281
Epoch 86/100
94/94 [==============================] - 0s 1ms/step - loss: 0.3749 - accuracy: 0.8308
Epoch 87/100
94/94 [==============================] - 0s 1ms/step - loss: 0.3730 - accuracy: 0.8258
Epoch 88/100
94/94 [==============================] - 0s 1ms/step - loss: 0.3737 - accuracy: 0.8231
Epoch 89/100
94/94 [==============================] - 0s 1ms/step - loss: 0.3789 - accuracy: 0.8227
Epoch 90/100
94/94 [==============================] - 0s 1ms/step - loss: 0.3718 - accuracy: 0.8318
Epoch 91/100
94/94 [==============================] - 0s 1ms/step - loss: 0.3700 - accuracy: 0.8284
Epoch 92/100
94/94 [==============================] - 0s 1ms/step - loss: 0.3703 - accuracy: 0.8268
Epoch 93/100


In [79]:
X_train, y_train = get_train_batch(df3_class0, df3_class1, 2990, 4130)

y_pred3 = ANN(X_train, y_train, X_test, y_test, 'binary_crossentropy', -1)

Epoch 1/100
83/83 [==============================] - 0s 1ms/step - loss: 0.6392 - accuracy: 0.6676
Epoch 2/100
83/83 [==============================] - 0s 1ms/step - loss: 0.5260 - accuracy: 0.7609
Epoch 3/100
83/83 [==============================] - 0s 1ms/step - loss: 0.4914 - accuracy: 0.7674
Epoch 4/100
83/83 [==============================] - 0s 1ms/step - loss: 0.4835 - accuracy: 0.7715
Epoch 5/100
83/83 [==============================] - 0s 1ms/step - loss: 0.4793 - accuracy: 0.7738
Epoch 6/100
83/83 [==============================] - 0s 1ms/step - loss: 0.4751 - accuracy: 0.7787
Epoch 7/100
83/83 [==============================] - 0s 1ms/step - loss: 0.4736 - accuracy: 0.7731
Epoch 8/100
83/83 [==============================] - 0s 1ms/step - loss: 0.4708 - accuracy: 0.7818
Epoch 9/100
83/83 [==============================] - 0s 1ms/step - loss: 0.4700 - accuracy: 0.7772
Epoch 10/100
83/83 [==============================] - 0s 1ms/step - loss: 0.4679 - accuracy: 0.7810
Epoch 11/

83/83 [==============================] - 0s 1ms/step - loss: 0.3709 - accuracy: 0.8357
Epoch 84/100
83/83 [==============================] - 0s 1ms/step - loss: 0.3704 - accuracy: 0.8387
Epoch 85/100
83/83 [==============================] - 0s 1ms/step - loss: 0.3687 - accuracy: 0.8391
Epoch 86/100
83/83 [==============================] - 0s 1ms/step - loss: 0.3675 - accuracy: 0.8444
Epoch 87/100
83/83 [==============================] - 0s 1ms/step - loss: 0.3649 - accuracy: 0.8410
Epoch 88/100
83/83 [==============================] - 0s 1ms/step - loss: 0.3669 - accuracy: 0.8440
Epoch 89/100
83/83 [==============================] - 0s 1ms/step - loss: 0.3677 - accuracy: 0.8387
Epoch 90/100
83/83 [==============================] - 0s 1ms/step - loss: 0.3633 - accuracy: 0.8459
Epoch 91/100
83/83 [==============================] - 0s 1ms/step - loss: 0.3626 - accuracy: 0.8463
Epoch 92/100
83/83 [==============================] - 0s 1ms/step - loss: 0.3614 - accuracy: 0.8433
Epoch 93/100


In [80]:
len(y_pred1)

1407

In [81]:
y_pred_final = y_pred1.copy()
for i in range(len(y_pred1)):
    n_ones = y_pred1[i] + y_pred2[i] + y_pred3[i]
    if n_ones>1:
        y_pred_final[i] = 1
    else:
        y_pred_final[i] = 0

In [82]:
cl_rep = classification_report(y_test, y_pred_final)
print(cl_rep)

              precision    recall  f1-score   support

           0       0.88      0.69      0.77      1033
           1       0.46      0.75      0.57       374

    accuracy                           0.70      1407
   macro avg       0.67      0.72      0.67      1407
weighted avg       0.77      0.70      0.72      1407



f1-score for minority class 1 improved to 0.62 from 0.57. The score for majority class 0 is suffering and reduced to 0.80 from 0.85 but at least there is some balance in terms of prediction accuracy across two classes